text/0000-metadata-for-offline-signers.md

```diff
@@ -0,0 +1,150 @@
+
+### Solution requirements
+
+#### Include metadata digest into signature
+
+Some cryptographically strong digest of metadata MUST be included into signable blob. There SHALL NOT be storage overhead for this blob, nor computational overhead, on the node side; thus MUST be a constant within given runtime version, deterministically defined by metadata.
+
+ - Metadata information that could be used in signable extrinsic decoding MUST be be included in digest;
+ - Digest MUST be deterministic with respect to metadata;
+ - Digest MUST be cryptographically strong against pre-image, both first and second;
+ - Extra-metadata information necessary for extrinsic decoding and constant within runtime version MUST be included in digest;
+ - Digest format SHOULD be versioned to allow rapid withdrawal of cold signing devices in case severe security vulnerability is found in shortener mechanism;
+ - Work necessary for proving metadata authenticity MAY be omitted at discretion of signer device design (to support automation tools).
+
+#### Reduce metadata size
+
+Metadata should be stripped from parts that are not necessary to parse a signable extrinsic, then it should be separated into a finite set of self-descriptive chunks. Thus, a subset of chunks necessary for signable extrinsic decoding and rendering could be sent, possibly in small portions (ultimately - one at a time), to cold device together with proof.
+
+ - Single chunk with proof payload size SHOULD fit within few kB;
+ - Chunks handling mechanism SHOULD support chunks being sent in any order without memory utilization overhead;
+ - Unused enum variants MUST be stripped (this has great impact on transmitted metadata size; examples: era enum, enum with all calls for call batching).
+
+## Stakeholders
+
+This feature is essential for **all** offline signer tools; many regular signing tools might make use of it. In general, this RFC greatly improves security of any network implementing it, as many governing keys are used with offline signers.
+
+Implementing this RFC would remove requirement to maintain metadata portals manually, as task of metadata verification would be effectively moved to consensus mechanism of the chain.
+
+## Explanation
+
+Detailed description of metadata shortening and digest process is provided in [metadata-shortener](https://github.com/Alzymologist/metadata-shortener) crate (see `cargo doc --open` and examples). Below are presented algorithms of the process.
+
+### Metadata descriptor
+
+Values for metadata shortening protocol version, `ExtrinsicMetadata`, SCALE-encoded `spec_version` and `spec_name` Strings, SCALE-encoded base58 prefix, SCALE-encoded decimals value, SCALE-encoded token unit String, should be prepared and combined as metadata descriptor.
```

Comment:
`base58prefix` is encoded as u16 and `decimals` as u8 in the reference implementation. It seems that fixed-width uints work better here since their size is known.
base58prefix is u16 from spec and using more than 255 decimal values is not rational since max value for u256 is 10e78.

Reply:
```suggestion
Values for metadata shortening protocol version, `ExtrinsicMetadata`, SCALE-encoded `spec_version` and `spec_name` Strings, `u16` base58 prefix, `u8` decimals value, SCALE-encoded token unit String, should be prepared and combined as metadata descriptor.
```